### PR TITLE
Wrap dimfilter

### PR DESCRIPTION
**Pull request by willschlitzer**
This pull request wraps the function `dimfilter`.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by meghanrjones** [comment]
Sorry for keeping you waiting on a review for this module! dimfilter will be a great addition to the PyGMT processing modules :tada: I am hoping to find time to check out the branch within the next week.

File: pygmt/src/dimfilter.py
Comment (willschlitzer) [reply pygmt/src/dimfilter.py]: I made this change in 4e7165f

File: pygmt/tests/test_dimfilter.py
Comment (willschlitzer) [reply pygmt/tests/test_dimfilter.py]: Updated to use `xarray` testing

File: pygmt/tests/test_dimfilter.py
Comment (willschlitzer) [reply pygmt/tests/test_dimfilter.py]: I changed the region to use asymmetric regions and different ranges.

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:2]: Is this short description accurate?

Comment (willschlitzer) [reply pygmt/src/dimfilter.py]: fixed in 87c7cb6; I must have left it in when copying another file

File: pygmt/tests/test_dimfilter.py
Comment (seisman) [inline pygmt/tests/test_dimfilter.py:21]: Please change the tests to use the `static_earth_relief` data.

Comment (willschlitzer) [reply pygmt/tests/test_dimfilter.py]: done in 8cb34e0

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:137]: Fix the linting issue:
```
************* Module pygmt.src.dimfilter
pygmt/src/dimfilter.py:137:30: C0201: Consider iterating the dictionary directly instead of calling .keys() (consider-iterating-dictionary)
```
```suggestion
                if "G" not in kwargs:  # if outgrid is unset, output to tempfile
```

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:130]:
```suggestion
            "The following parameters are required: distance, filters, sectors."
```

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:128]: https://github.com/GenericMappingTools/pygmt/blob/c306352a0c9f965c0fd52cc4f46f1a6074bb58b4/pygmt/helpers/utils.py#L260

We have the `args_in_kwargs` function to simplify line 128.

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:131]:
```suggestion
        raise GMTInvalidInput(
            """At least one of the following parameters must be specified:
            distance, filters, sectors."""
        )
```

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:115]:
```suggestion
        [*xmin*, *xmax*, *ymin*, *ymax*].
```

File: pygmt/tests/test_dimfilter.py
Comment (weiji14) [inline pygmt/tests/test_dimfilter.py:62]:
```suggestion
def test_dimfilter_no_outgrid(grid, expected_grid):
```

File: pygmt/src/dimfilter.py
Comment (willschlitzer) [reply pygmt/src/dimfilter.py:128]: Unless I misunderstand its function, isn't `args_in_kwargs` to test if _any_ of the arguments are in the kwargs? This is making sure that all 3 arguments are in the kwargs.

Comment (seisman) [reply pygmt/src/dimfilter.py:128]: There must be some misunderstanding here. The exception message says "At least one of the following parameters must be specified: distance, filters, or sectors.", but `if ("D" not in kwargs) or ("F" not in kwargs) or ("N" not in kwargs):` requires all the three are in the `kwargs`.

Reading the dimfilter documentation (https://docs.generic-mapping-tools.org/dev/dimfilter), a command like `gmt dimfilter depths.txt -Q > output.z` is also valid.

Comment (weiji14) [reply pygmt/src/dimfilter.py:128]: So the check should probably look something like this:

```suggestion
    if not all(arg in kwargs for arg in ["D", "F", "N"]) and "Q" not in kwargs:
```

However, the `-Q` mode requires a table input, and that has not been properly implemented in this Pull Request yet. Do you think we should just have the `grid`->`grid` functionality finalized, and then do `table`->`grid` in a separate PR?

File: pygmt/src/dimfilter.py
Comment (weiji14) [inline pygmt/src/dimfilter.py:136]: Since `dimfilter` accepts both vector or raster, no need to check for it here.

```suggestion
            file_context = lib.virtualfile_from_data(check_kind=None, data=grid)
```

File: pygmt/tests/test_dimfilter.py
Comment (weiji14) [inline pygmt/tests/test_dimfilter.py:78]:
```suggestion
    Check that dimfilter fails correctly when not all of sectors, filters, and
    distance are specified.
```

File: pygmt/tests/test_dimfilter.py
Comment (weiji14) [inline pygmt/tests/test_dimfilter.py:81]:
```suggestion
        dimfilter(grid=grid, sectors="l6", distance=4)
```

File: pygmt/src/dimfilter.py
Comment (willschlitzer) [reply pygmt/src/dimfilter.py:128]: @weiji14 I would prefer to get this PR wrapped up, rather than try and work out table to grid functionality. I'm assuming that means leaving out the mention of `Q` in this if statement?

File: pygmt/src/dimfilter.py
Comment (seisman) [inline pygmt/src/dimfilter.py:27]:
```suggestion
@kwargs_to_strings(I="sequence", R="sequence")
```

File: pygmt/src/dimfilter.py
Comment (weiji14) [inline pygmt/src/dimfilter.py:77]:
```suggestion
        **x**\ *width*\ [**+l**\|\ **u**].
```

File: pygmt/src/dimfilter.py
Comment (weiji14) [inline pygmt/src/dimfilter.py:79]: Append doesn't quite make sense in PyGMT, maybe just say 'Use'?
```suggestion
        non-convolution filters. Use the filter code **x** followed by
```

File: pygmt/src/dimfilter.py
Comment (weiji14) [inline pygmt/src/dimfilter.py:108]:
```suggestion
    spacing : str or list
```

**Review by weiji14 — COMMENTED** [review]
Just some minor docstring comments. The implementation looks quite good already, and we could merge this new feature 'soon' after a final review, but it'll depend on whether we want to do a PyGMT v0.6.1 or jump straight to PyGMT v0.7.0.

I say we wait until the end of March 2022, and if there isn't any major bug that pops up with users using PyGMT v0.6.0, then we'll jump straight to v0.7.0. What do you think?

**Comment by willschlitzer** [comment]
> I say we wait until the end of March 2022, and if there isn't any major bug that pops up with users using PyGMT v0.6.0, then we'll jump straight to v0.7.0. What do you think?

That's fine with me; I'm just hoping to get this wrapped up relatively soon since it's been an open for 6 months.

**Review by seisman — APPROVED** [review]

**Comment by seisman** [comment]
> I say we wait until the end of March 2022, and if there isn't any major bug that pops up with users using PyGMT v0.6.0, then we'll jump straight to v0.7.0. What do you think?

I agree.

**Comment by willschlitzer** [comment]
@GenericMappingTools/pygmt-maintainers  Are we still holding off for v0.6.1 or am I good to merge this?

**Comment by seisman** [comment]
We will release v0.6.1 soon https://github.com/GenericMappingTools/pygmt/issues/1856.

**Pull request merged**
